internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,248 @@
+//
+// • logger must be provided and is used to log errors that occur during scheduling and running of jobs
+//
+// WithRunJobsLimit and WithRunJobsInterval are the only valid options.
+func New(serverId string, jobRepoFn jobRepo.JobRepoFactory, logger hclog.Logger, opt ...Option) (*Scheduler, error) {
+	const op = "scheduler.New"
+	if serverId == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing server id")
+	}
+	if jobRepoFn == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing job repo function")
+	}
+	if logger == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing logger")
+	}
+
+	opts := getOpts(opt...)
+	return &Scheduler{
+		serverId:        serverId,
+		jobRepoFn:       jobRepoFn,
+		logger:          logger,
+		registeredJobs:  new(sync.Map),
+		runningJobs:     make(map[JobId]runningJob),
+		runJobsLimit:    opts.withRunJobsLimit,
+		runJobsInterval: opts.withRunJobInterval,
+	}, nil
+}
+
+// RegisterJob registers a job with the scheduler and persists the job into the repository.
+//
+// • job must be provided and is an implementer of the Job interface.
+//
+// • code must be provided and is used to uniquely distinguish between the same job implementation.
+//
+// The returned JobId is unique to the provide job and code.
+// All options are ignored.
+func (s *Scheduler) RegisterJob(ctx context.Context, job Job, code string, _ ...Option) (JobId, error) {
+	const op = "scheduler.(Scheduler).RegisterJob"
+	if err := validateJob(job); err != nil {
+		return "", errors.Wrap(err, op)
+	}
+	if code == "" {
+		return "", errors.New(errors.InvalidParameter, op, "missing code")
+	}
+
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return "", errors.Wrap(err, op)
+	}
+
+	j, err := repo.CreateJob(ctx, job.Name(), code, job.Description())
+	if err != nil {
+		return "", errors.Wrap(err, op)
+	}
+	s.registeredJobs.Store(JobId(j.PrivateId), job)
+
+	return JobId(j.PrivateId), nil
+}
+
+// UpdateJobNextRun sets the next scheduled run time for the provided jobId
+// to the current database time incremented by the nextRunIn parameter.  If
+// nextRunIn == 0 the job will be available to run immediately.
+//
+// • jobId must be provided and is the id of the job returned by RegisterJob
+//
+// All options are ignored.
+func (s *Scheduler) UpdateJobNextRun(ctx context.Context, jobId JobId, nextRunIn time.Duration, _ ...Option) error {
+	const op = "scheduler.(Scheduler).UpdateJobNextRun"
+	if jobId == "" {
+		return errors.New(errors.InvalidParameter, op, "missing job id")
+	}
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	_, err = repo.UpdateJobNextRun(ctx, string(jobId), nextRunIn)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+	return nil
+}
+
+// Start begins the scheduling loop that will query the repository for jobs to run and
+// run them in a goroutine.
+func (s *Scheduler) Start() {
+	if s.started.Load() {
```

Comment:
Shouldn't this be an atomic [compare-and-swap](https://pkg.go.dev/go.uber.org/atomic#Bool.CAS) operation?